metric-publishers/cloudwatch-metric-publisher/src/main/java/software/amazon/awssdk/metrics/publishers/cloudwatch/internal/transform/TimeBucketedMetrics.java

```diff
@@ -0,0 +1,226 @@
+        return metrics.creationTime().truncatedTo(MINUTES);
+    }
+
+    private void addMetricsToBucket(MetricCollection metrics, Instant bucketId) {
+        aggregateMetrics(metrics, timeBucketedMetrics.computeIfAbsent(bucketId, i -> new HashMap<>()));
+    }
+
+    private void aggregateMetrics(MetricCollection metrics, Map<MetricAggregatorKey, MetricAggregator> bucket) {
+        List<Dimension> dimensions = dimensions(metrics);
+        extractAllMetrics(metrics).forEach(metricRecord -> {
+            MetricAggregatorKey aggregatorKey = new MetricAggregatorKey(metricRecord.metric(), dimensions);
+            valueFor(metricRecord).ifPresent(metricValue -> {
+                bucket.computeIfAbsent(aggregatorKey, m -> newAggregator(aggregatorKey))
+                      .addMetricValue(MetricValueNormalizer.normalize(metricValue));
+            });
+        });
+    }
+
+    private List<Dimension> dimensions(MetricCollection metricCollection) {
+        List<Dimension> result = new ArrayList<>();
+        for (MetricRecord<?> metricRecord : metricCollection) {
+            if (dimensions.contains(metricRecord.metric())) {
+                result.add(Dimension.builder()
+                                    .name(metricRecord.metric().name())
+                                    .value((String) metricRecord.value())
+                                    .build());
+            }
+        }
+
+        // Sort the dimensions to make sure that the order in the input metric collection doesn't affect the result.
+        // We use descending order just so that "ServiceName" is before "OperationName" when we use the default dimensions.
+        result.sort(Comparator.comparing(Dimension::name).reversed());
+        return result;
+    }
+
+    private List<MetricRecord<?>> extractAllMetrics(MetricCollection metrics) {
+        List<MetricRecord<?>> result = new ArrayList<>();
+        extractAllMetrics(metrics, result);
+        return result;
+    }
+
+    private void extractAllMetrics(MetricCollection metrics, List<MetricRecord<?>> extractedMetrics) {
+        for (MetricRecord<?> metric : metrics) {
+            extractedMetrics.add(metric);
+        }
+        metrics.children().forEach(child -> extractAllMetrics(child, extractedMetrics));
+    }
+
+    private MetricAggregator newAggregator(MetricAggregatorKey aggregatorKey) {
+        SdkMetric<?> metric = aggregatorKey.metric();
+        StandardUnit metricUnit = unitFor(metric);
+        if (detailedMetrics.contains(metric)) {
+            return new DetailedMetricAggregator(aggregatorKey, metricUnit);
+        } else {
+            return new SummaryMetricAggregator(aggregatorKey, metricUnit);
+        }
+    }
+
+    private StandardUnit unitFor(SdkMetric<?> metric) {
+        Class<?> metricType = metric.valueClass();
+
+        if (Duration.class.isAssignableFrom(metricType)) {
+            return StandardUnit.MILLISECONDS;
+        }
```

Comment:
I noticed we updated V1 from milliseconds to nanoseconds. Should we just do nanos instead from the beginning for V2?